src/lib/redirects-manager-context.tsx

```diff
@@ -0,0 +1,116 @@
+import * as React from 'react';
+import { v4 } from 'uuid';
+import { saveState, checkRepeatedRequests } from './utils';
+
+export interface RedirectsManagerContextInterface {
+    store: RedirectionsStore;
+    wildcard: boolean;
+    saving?: boolean;
+    lastSave?: Date;
+    lastModification?: Date;
+    filterBy?: string;
+    perPage?: number;
+    currentPage?: number;
+    imported?: boolean;
+}
+
+export interface RedirectionProps {
+    id: string;
+    request?: string;
+    destination?: string;
+    modificationDate?: any;
+    warningRequestDuplication?: boolean;
+}
+
+export interface RedirectionsStore extends Array<RedirectionProps> { }
+
+type Dispatch = (action: Action) => void;
+interface RedirectsManagerProviderProps { children: React.ReactNode; }
+
+const RedirectsManagerContext = React.createContext<RedirectsManagerContextInterface | undefined>(undefined);
+const RedirectsManagerDispatchContext = React.createContext<Dispatch | undefined>(undefined);
+
+interface UpdateServerStateProps {dispatch: Dispatch; state: RedirectsManagerContextInterface; }
+
+export const updateServerState = async ({dispatch, state}: UpdateServerStateProps) => {
+    dispatch({type: 'saving-state', value: true});
+    try {
+        const newState = await saveState(state);
+        dispatch({type: 'set', value: {...newState, lastSave: new Date(), saving: false}});
+    } catch (e) {
+        dispatch({type: 'saving-state', value: false});
+    }
+};
+
+type Action = {type: 'saving-state', value: boolean} |
+              {type: 'add', value: null} |
+              {type: 'remove', value: RedirectionProps} |
+              {type: 'edit', value: RedirectionProps} |
+              {type: 'set', value: RedirectsManagerContextInterface};
+
+const redirectsManagerReduducer = (state: RedirectsManagerContextInterface, action: Action) => {
+    switch (action.type) {
+
+        case 'add': {
+            state.store.push({id: v4(), modificationDate: 'not saved'});
+            return {...state, currentPage: 1};
+        }
+
+        case 'edit': {
+            const index = state.store.findIndex((el) => {
+                return el.id === action.value.id;
+            });
+            state.store[index] = action.value;
```

Comment:
please account for out-of-range issues, i.e. `findIndex` may return `-1` if all tests fail

Reply:
In this scenario it was not likely, as the reducer action is triggered only from the respective redirection that is passing its value back, anyway, I did account it in extreme precaution. Also made the returning of the newState explicit